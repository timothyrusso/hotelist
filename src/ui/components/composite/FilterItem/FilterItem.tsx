import { FC } from "react";
import { View } from "react-native";
import { FilterChip } from "../../basic/FilterChip/FilterChip";
import {
	FilterTitle,
	type FontistoName,
} from "../../basic/FilterTitle/FilterTitle";
import { styles } from "./FilterItem.style";

type FilterItemProps = {
	icon: FontistoName;
	title: string;
	data: string[] | number[];
	showIcon?: boolean;
	onPress: (value: string | number) => void;
	selectedData: number[] | string[];
};

export const FilterItem: FC<FilterItemProps> = ({
	icon,
	title,
	data,
	showIcon = false,
	onPress,
	selectedData,
}) => {
	return (
		<View style={styles.container}>
			<FilterTitle icon={icon} title={title} />
			<View style={styles.dataContainer}>
				{data.map((item) => (
					<FilterChip
						key={item}
						value={item}
						icon={showIcon ? icon : undefined}
						onPress={onPress}
						selectedData={selectedData}
					/>
				))}
			</View>
		</View>
	);
};
