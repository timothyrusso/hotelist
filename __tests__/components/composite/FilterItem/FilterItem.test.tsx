import { FilterItem } from "@/src/ui/components/composite/FilterItem/FilterItem";
import { fireEvent, render } from "@testing-library/react-native";

// Mock child components to isolate FilterItem testing
jest.mock("@/src/ui/components/basic/FilterTitle/FilterTitle", () => {
	// eslint-disable-next-line @typescript-eslint/no-require-imports
	const { View, Text } = require("react-native");
	return {
		FilterTitle: ({ icon, title }: { icon: string; title: string }) => (
			<View testID="filter-title">
				<Text>
					{icon} - {title}
				</Text>
			</View>
		),
	};
});

jest.mock("@/src/ui/components/basic/FilterChip/FilterChip", () => {
	// eslint-disable-next-line @typescript-eslint/no-require-imports
	const { Pressable, Text } = require("react-native");
	return {
		FilterChip: ({
			value,
			icon,
			onPress,
			selectedData,
		}: {
			value: string | number;
			icon?: string;
			onPress: (value: string | number) => void;
			selectedData: (string | number)[];
		}) => (
			<Pressable testID={`filter-chip-${value}`} onPress={() => onPress(value)}>
				<Text testID={`chip-selected-${value}`}>
					{selectedData.includes(value) ? "selected" : "not-selected"}
				</Text>
				<Text>{icon ? `${icon}-${value}` : value}</Text>
			</Pressable>
		),
	};
});

describe("FilterItem", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	describe("snapshot tests", () => {
		it("should render correctly with default props", () => {
			// Mock data and callback
			const mockData = ["Option 1", "Option 2", "Option 3"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { toJSON } = render(
				<FilterItem
					icon="star"
					title="Test Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(toJSON()).toMatchSnapshot();
		});

		it("should render correctly with showIcon enabled", () => {
			const mockData = [1, 2, 3, 4, 5];
			const mockOnPress = jest.fn();
			const mockSelectedData: number[] = [2, 4];

			const { toJSON } = render(
				<FilterItem
					icon="star"
					title="Star Rating"
					data={mockData}
					showIcon={true}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(toJSON()).toMatchSnapshot();
		});
	});

	describe("functional tests", () => {
		it("should render FilterTitle with correct props", () => {
			const mockData = ["Option 1", "Option 2"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { getByTestId } = render(
				<FilterItem
					icon="star"
					title="Test Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(getByTestId("filter-title")).toBeTruthy();
		});

		it("should render FilterChip for each data item", () => {
			const mockData = ["Option 1", "Option 2", "Option 3"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { getByTestId } = render(
				<FilterItem
					icon="star"
					title="Test Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(getByTestId("filter-chip-Option 1")).toBeTruthy();
			expect(getByTestId("filter-chip-Option 2")).toBeTruthy();
			expect(getByTestId("filter-chip-Option 3")).toBeTruthy();
		});

		it("should not pass icon to FilterChip when showIcon is false", () => {
			const mockData = ["Option 1", "Option 2"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			// Render the component with showIcon disabled (default)
			const { getByTestId } = render(
				<FilterItem
					icon="star"
					title="Test Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			// Verify FilterChip does not receive icon prop
			const chip1 = getByTestId("filter-chip-Option 1");
			expect(chip1).toBeTruthy();
		});

		it("should propagate onPress callback to FilterChip", () => {
			const mockData = ["Option 1", "Option 2"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { getByTestId } = render(
				<FilterItem
					icon="star"
					title="Test Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			// Press the first chip
			fireEvent.press(getByTestId("filter-chip-Option 1"));

			// Verify onPress was called with correct value
			expect(mockOnPress).toHaveBeenCalledWith("Option 1");
		});
	});

	describe("edge cases", () => {
		it("should handle empty data array", () => {
			const mockData: string[] = [];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { toJSON } = render(
				<FilterItem
					icon="star"
					title="Empty Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(toJSON()).toBeTruthy();
		});

		it("should handle single item in data array", () => {
			const mockData = ["Single Option"];
			const mockOnPress = jest.fn();
			const mockSelectedData: string[] = [];

			const { getByTestId } = render(
				<FilterItem
					icon="star"
					title="Single Filter"
					data={mockData}
					onPress={mockOnPress}
					selectedData={mockSelectedData}
				/>,
			);

			expect(getByTestId("filter-chip-Single Option")).toBeTruthy();
		});
	});
});
