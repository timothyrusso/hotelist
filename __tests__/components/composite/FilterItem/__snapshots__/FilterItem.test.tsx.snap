// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterItem snapshot tests should render correctly with default props 1`] = `
<View
  style={
    {
      "borderBottomColor": "#e8e8e8",
      "borderBottomWidth": 1,
      "marginHorizontal": 20,
      "paddingVertical": 24,
      "rowGap": 16,
    }
  }
>
  <View
    testID="filter-title"
  >
    <Text>
      star
       - 
      Test Filter
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 12,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-Option 1"
    >
      <Text
        testID="chip-selected-Option 1"
      >
        not-selected
      </Text>
      <Text>
        Option 1
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-Option 2"
    >
      <Text
        testID="chip-selected-Option 2"
      >
        not-selected
      </Text>
      <Text>
        Option 2
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-Option 3"
    >
      <Text
        testID="chip-selected-Option 3"
      >
        not-selected
      </Text>
      <Text>
        Option 3
      </Text>
    </View>
  </View>
</View>
`;

exports[`FilterItem snapshot tests should render correctly with showIcon enabled 1`] = `
<View
  style={
    {
      "borderBottomColor": "#e8e8e8",
      "borderBottomWidth": 1,
      "marginHorizontal": 20,
      "paddingVertical": 24,
      "rowGap": 16,
    }
  }
>
  <View
    testID="filter-title"
  >
    <Text>
      star
       - 
      Star Rating
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 12,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-1"
    >
      <Text
        testID="chip-selected-1"
      >
        not-selected
      </Text>
      <Text>
        star-1
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-2"
    >
      <Text
        testID="chip-selected-2"
      >
        selected
      </Text>
      <Text>
        star-2
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-3"
    >
      <Text
        testID="chip-selected-3"
      >
        not-selected
      </Text>
      <Text>
        star-3
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-4"
    >
      <Text
        testID="chip-selected-4"
      >
        selected
      </Text>
      <Text>
        star-4
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="filter-chip-5"
    >
      <Text
        testID="chip-selected-5"
      >
        not-selected
      </Text>
      <Text>
        star-5
      </Text>
    </View>
  </View>
</View>
`;
